### PR TITLE
Feature queries in globe view

### PR DESCRIPTION
**Pull request by mpulkki-mapbox**
## Technical introduction
Feature queries can be used to identify which features are visible at specified locations. In gl-js this functionality can be accessed with `map.queryRenderedFeatures`.

From a technical point of view the logic of finding rendered features on the screen is a collection intersection tests between the query geometry and polygons of features as they would appear on the screen. A naive approach would be to go through all of the features in visible tiles to see if they're intersecting with the query geometry, but a more sophisticated logic is used to reduce computations required in order to achieve better performance. The process of finding intersecting features can be divided into two phases: **coarse** and **precise**. The idea is to run more coarse (and hence cheaper) intersection tests to reduce the set of potential features to as small as possible before performing the more precise and expensive tests to individual features.

The query begins with definition of two query rectangles: the first one is the query rectangle itself that's either a screen rectangle passed to the function by the caller or a "point rectangle" that has no volume. Additionally a second rectangle is constructed by adding a predefined pixel padding to the query rectangle. Both of these are then projected to mercator coordinates (and called "query polygons" from this point onwards). Intersection tests can then begin:

#### Coarse phase
 1. First step in the coarse phase is to find potentially visible tiles from each loaded source by checking whether the *padded query polygon* intersects with tile boundaries in tile space. A slightly modified polygon is used if the style contains 3D layers (e.g. fill extrusions) as extruded geometries might intersect with the query polygon even though tiles themselves might not. In these cases the query rectangle in screen coordinates is extended by adding screen coordinate of the camera location on the surface and is then projected to mercator again.
 2. Next step of the broad phase is done for tiles that passed the previous step. A specific per-tile acceleration structure "FeatureIndex" that splits the tile into a 16x16 grid is used to narrow potentially queryable features inside a single tile even further. Again the *padded query polygon* is used to find unique features assigned to intersecting cells of the gride. The precise phase can finally begin after potential features have been collected.

#### Precise phase
In the precise phase all features fetched from the feature index are checked with layer specific logic. Depending on the alignment of the owning layer (map or viewport) features are either projected to the screen as they would be rendered or kept in tile space. Viewport aligned feature are projected to the screen and checked against the original *screen space query rectangle* whereas map aligned features are checked against the *non-padded query polygon* in tile coordinates.

Following layers can be queried for features (based on query logic implementations I found):
 - Circle, Fill, Fill extrusion, Heatmap, Line and Symbol*
Layers not queryable
 - Background, Custom, Hillshade, Raster, Sky

*Symbol layer querying is done purely using the collision index for both map and viewport aligned symbols.

## Feature queries on globe

Adding feature query support to globe is a *quite* straightforward task as most of the intersection logic is done in tile space. Conversion from screen coordinates to tile coordinates is a well defined operation also in the globe view so the only areas requiring update are mostly related to transformation logic where globe matrices are used. 

There are some challenges in this approach though. The first one being is that the projection from screen to mercator in globe view is a non-linear operation, i.e. straight lines in the world space are not straight when mapped on the surface of the globe. What this means is that the query polygon created by projecting screen space rectangle does not represent the actual area when viewed in mercator units (it's the same effect when flight paths do not appear straight in mercator projection.) Another challenge is to support queries during transition between mercator and globe projections.

Following videos visualizes the naive projection of screen space rectangles on the map:

https://user-images.githubusercontent.com/55925868/161548250-d8c15ee1-4595-4c70-a9dc-6d843e037d7f.mp4

https://user-images.githubusercontent.com/55925868/161548348-616e8b1c-6a4d-4cf6-9acd-1cfa663fd064.mp4

Changes introduced in this PR can be split into two parts:
 1. The existing logic of using tile space polygons is fully preserved. As an additional step query rectangles are "resampled", ie. they're projected to mercator coordinates by adding enough intermediate points to the edges of the polygon so that the lines will become correctly curved. This will have some performance implications especially on lower zoom levels as query polygons are more complex.
 2. Each layer type supporting feature queries have been updated to properly support globe transformations. In practice this means that viewport aligned features and 3D layers (such as fill extrusions) have been updated.

I've also updated following debug pages to support globe view `debug/featurestate.html` and `debug/query_features.html` and added a new one `debug/globe-feature-query.html`.

https://user-images.githubusercontent.com/55925868/161912523-b426ac5d-3b22-423a-85d7-1d643b120d96.mp4

What is not included are left for follow-up PRs. This includes:
 1. Support queries during globe <-> merc transition
 2. ~~Query results are not accurate when either south or north pole is inside the query polygon. This is because the projection is non-trivial in these cases.~~
 
     https://user-images.githubusercontent.com/55925868/161913111-9769ede5-134b-416c-81a9-fe673530ac3d.mp4
 
 3. Queries on features near horizon might not be fully accurate. This is because we're rendering features even they're slightly behind the horizon (ie. occluded by the ground) but the query polygon doesn't extend that far.

     https://user-images.githubusercontent.com/55925868/161914815-b903213b-4b03-4911-bb67-6781a7429450.mp4

 5. Performance profiling and optimization

## Launch Checklist

 - [x] briefly describe the changes in this PR
 - [x] include before/after visuals or gifs if this PR includes visual changes
 - [x] write tests for all new functionality
 - [x] manually test the debug page
 - [x] apply changelog label ('bug', 'feature', 'docs', etc) or use the label 'skip changelog'

### CODE REVIEW

**Comment by mpulkki-mapbox** [comment]
I ended up including the support for feature queries covering south and north pole in this pull request. As mentioned in the opening ticket, projecting a screen rectangle over a pole is a bit more involving task. A naive approach with the existing logic would result in following looking projection as only corner points are projected:

initial_screen_rect
initial projection

How I ended up implementing this is described in the following steps:
 1. Find if either north or south pole is inside the query rectangle by projecting their world positions to the screen. The code path is slightly different depending on which pole is inside the query area.
 2. Project all 4 corner points to Mercator coordinates and find which of the 4 edges is crossing the antimeridian.
 3. Project the 3 other edges with resampling to Mercator coordinates forming a multi point line. This will be the starting point of the resulting query polygon.
 4. Take the edge crossing the antimeridian and split it into two parts at the intersection location. With these two segments added to the polygon we now have the initial screen rectangle projected into a continuous line in Mercator coordinates.
 5. The final step is to connect both ends of the line by encircling the corresponding pole.

anim_0
anim_1

**Comment by astojilj** [comment]
> How I ended up implementing this is described in the following steps:
> 
...
> 
> anim_0 anim_1

@mpulkki-mapbox 
 This documentation and animation is amazing. Code comments are clarifying this well too.

**Comment by astojilj** [comment]
Verified using [cluster](https://docs.mapbox.com/mapbox-gl-js/example/cluster/) and [bounding box](https://docs.mapbox.com/mapbox-gl-js/example/using-box-queryrenderedfeatures/) examples. Works great.

https://user-images.githubusercontent.com/549216/162734649-944c2edd-733c-4662-9132-3e2aa373bc61.mov

File: src/data/bucket/fill_extrusion_bucket.js [hunk 1]
```diff
@@ -605,3 +587,39 @@ function tileToMeter(canonical: CanonicalTileID) {
     // simplify cos(2 * atan(e) - PI/2) from mercator_coordinate.js, remove trigonometrics.
     return circumferenceAtEquator * 2 * exp / (exp * exp + 1) / EXTENT / (1 << canonical.z);
 }
+
+export function fillExtrusionHeightLift(): number {
+    // A rectangle covering globe is subdivided into a grid of 32 cells
+    // This information can be used to deduce a minimum lift value so that
+    // fill extrusions with 0 height will never go below the ground.
+    const angle = Math.PI / 32.0;
```
[the remark this answers is not in the document]
Comment (astojilj) [reply src/data/bucket/fill_extrusion_bucket.js:595]: Got it elsewhere... full globe is 2 * PI?

File: src/geo/projection/globe.js [hunk 2]
```diff
@@ -164,6 +164,11 @@ export default class Globe extends Mercator {
         return new MercatorCoordinate(mx, my);
     }
 
+    pointCoordinate3D(tr: Transform, x: number, y: number): ?Vec4 {
+        const mc = this.pointCoordinate(tr, x, y, 0);
```
Comment (astojilj) [inline src/geo/projection/globe.js:168]: Logging offline discussion:
Terrain.pointCoordinate raycasting is implemented only for Mercator projection and is acceptable implementation constraint.

File: src/geo/transform.js [hunk 3]
```diff
@@ -1227,8 +1228,7 @@ class Transform {
      */
     pointCoordinate3D(p: Point): MercatorCoordinate {
         if (!this.elevation) return this.pointCoordinate(p);
```
Comment (astojilj) [inline src/geo/transform.js:1230]: Globe.pointCoordinate seem to already clamp to the closest point on the sphere. Is this just preparation for after Terrain.pointCoordinate gets implemented for globe projection? And currently, iterative approach is not needed for globe?

File: src/util/util.js [hunk 4]
```diff
@@ -139,7 +139,7 @@ export function polygonizeBounds(min: Point, max: Point, buffer: number = 0, clo
     const polygon = [minBuf, new Point(maxBuf.x, minBuf.y), maxBuf, new Point(minBuf.x, maxBuf.y)];
 
     if (close) {
-        polygon.push(minBuf);
+        polygon.push(minBuf.clone());
```
Comment (astojilj) [inline src/util/util.js:142]: What issue did the clone resolve?

**Review by astojilj — COMMENTED** [review]
LGTM. Have several minor questions, will do another pass after answers or potential code changes.

File: src/style/query_geometry.js [hunk 5]
```diff
@@ -217,27 +281,244 @@ export class QueryGeometry {
...
+        if (direction < 0) {
+            if (a < b) {
+                return {idx: i, t: -a / (b - 1 - a)};
+            }
+        } else {
+            if (b < a) {
+                return {idx: i, t: (1 - a) / (b + 1 - a)};
+            }
+        }
+    }
```
Comment (karimnaaji) [inline src/style/query_geometry.js:519]: We search for one intersection, but how are polygons crossing the antimeridian several times handled?

**Review by karimnaaji — COMMENTED** [review]
Looks good to me and works great! I will do more manual testing in a second pass. Otherwise, I have a few comments asking for clarification, but nothing major so far.

File: src/data/bucket/fill_extrusion_bucket.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (mpulkki-mapbox) [reply src/data/bucket/fill_extrusion_bucket.js:595]: This piece of code was indeed moved as it's required in querying of fill extrusions

File: src/geo/transform.js (hunk 3 above)
Comment (mpulkki-mapbox) [reply src/geo/transform.js:1230]: I think the original idea for the clamping behavior was to keep the function signature intact. There's no option for returning a null value if nothing was picked. This was also the easiest way forward as many parts of the codebase (such as feature querying) relies on this kind of behavior.

But this is a really good question to ask! Both the function `projection.pointCoordinate` and this `transform.pointCoordinate3D` should be revisited at some point.

File: src/util/util.js (hunk 4 above)
Comment (mpulkki-mapbox) [reply src/util/util.js:142]: I spent a great deal of time hunting a bug that was caused by a one point in the polygon being a "shared reference". I don't think it makes sense to have such points as it doesn't allow in-place modifications to the points in the array. For example the `resample` function would not work.

File: src/style/query_geometry.js [hunk 6]
```diff
@@ -217,27 +281,244 @@ export class QueryGeometry {
...
-    _bufferedScreenMercator(padding: number, transform: Transform): MercatorCoordinate[] {
+    _bufferedScreenMercator(padding: number, transform: Transform): CachedPolygon {
         const key = cacheKey(padding);
         if (this._screenRaycastCache[key]) {
             return this._screenRaycastCache[key];
         } else {
-            const poly = this.bufferedScreenGeometry(padding).map((p) => transform.pointCoordinate3D(p));
+            let poly: CachedPolygon;
+
+            if (transform.projection.name === 'globe') {
```
[the remark this answers is not in the document]
Comment (mpulkki-mapbox) [reply src/style/query_geometry.js:291]: I do agree. Some kind of enumeration system would be preferable here. This function is called only once or twice per feature query though so it's not particularly hot path.

File: src/style/query_geometry.js [hunk 7]
```diff
@@ -217,27 +281,244 @@ export class QueryGeometry {
...
+    }
+
+    return {
+        polygon: resampled.map(p => new MercatorCoordinate(p.x, p.y)),
+        unwrapped: false
+    };
+}
+
+function resamplePolygon(polygon: Point[], transform: Transform): Point[] {
+    const tolerance = 1.0 / 256.0;
```
[the remark this answers is not in the document]
Comment (mpulkki-mapbox) [reply src/style/query_geometry.js:488]: Good point, I'll add a comment :+1:

File: src/geo/projection/projection.js [hunk 8]
```diff
@@ -92,6 +92,16 @@ export default class Projection {
         return tr.rayIntersectionCoordinate(tr.pointRayIntersection(clamped, z));
     }
 
+    pointCoordinate3D(tr: Transform, x: number, y: number): ?Vec4 {
+        const p = new Point(x, y);
+        if (tr.elevation) {
+            return tr.elevation.pointCoordinate(p);
+        } else {
+            const mc = this.pointCoordinate(tr, p.x, p.y, 0);
+            return [mc.x, mc.y, mc.z, mc.toAltitude()];
```
[the remark this answers is not in the document]
Comment (mpulkki-mapbox) [reply src/geo/projection/projection.js:101]: That's a good question. I believe that's a leftover and I copied it without thinking too much. I've updated the code to return ?Vec3

File: src/geo/projection/projection.js [hunk 9]
```diff
@@ -92,6 +92,16 @@ export default class Projection {
         return tr.rayIntersectionCoordinate(tr.pointRayIntersection(clamped, z));
     }
 
+    pointCoordinate3D(tr: Transform, x: number, y: number): ?Vec4 {
+        const p = new Point(x, y);
+        if (tr.elevation) {
+            return tr.elevation.pointCoordinate(p);
```
[the remark this answers is not in the document]
Comment (mpulkki-mapbox) [reply src/geo/projection/projection.js:98]: Good find! I don't know why flow check didn't catch that one. I've modified the function signature to return ?Vec3

**Review by astojilj — APPROVED** [review]

**Review by karimnaaji — APPROVED** [review]
LGTM @mpulkki-mapbox , nice work! The two issues I noticed from testing this feature:

**1. Reduced precision around the horizon**

https://user-images.githubusercontent.com/7061573/163066508-1b20852a-e96d-433b-817c-e95b354d7c0e.mov

**2. False positive around horizon due to horizon clamping**

https://user-images.githubusercontent.com/7061573/163066618-db297271-4c5e-4983-877d-aad92f9e1442.mov

The second issue is fairly easy to trigger and should be captured as a follow up fix, and I think you already are aware of the first one. Could you capture these two follow-up items in issues?

**Pull request merged**